File: android/src/main/kotlin/my/alexl/multicamera/Registry.kt
package my.alexl.multicamera

import android.media.Image
import android.os.Handler
import android.os.Looper

class Registry(val plugin: MulticameraPlugin) {
    val cameras = HashMap<Long, Camera>()
    val cameraHandles = HashMap<Camera.Direction, CameraHandle>()

    fun registerCamera(
        direction: Camera.Direction,
        paused: Boolean,
        recognizeText: Boolean,
        scanBarcodes: Boolean,
        detectFaces: Boolean
    ): Long {
        val camera = Camera(
            plugin,
            direction,
            paused,
            recognizeText,
            scanBarcodes,
            detectFaces
        )
        cameras[camera.id] = camera
        reconcile()

        return camera.id
    }

    fun updateCamera(
        id: Long,
        direction: Camera.Direction,
        paused: Boolean,
        recognizeText: Boolean,
        scanBarcodes: Boolean,
        detectFaces: Boolean
    ) {
        cameras[id]?.let {
            it.direction = direction
            it.paused = paused
            it.recognizeText = recognizeText
            it.scanBarcodes = scanBarcodes
            it.detectFaces = detectFaces
            reconcile()
        }
    }

    fun unregisterCamera(id: Long) {
        cameras.remove(id)?.close()
        reconcile()
    }

    fun captureImage(id: Long, callback: (ByteArray?) -> Unit) {
        val camera = cameras[id]
        if (camera == null) {
            callback(null)
            return
        }

        val handle = cameraHandles[camera.direction]
        if (handle == null) {
            callback(null)
            return
        }

        handle.captureImage(callback)
    }

    fun onOrientationChanged() {
        for (handle in cameraHandles.values) {
            handle.updateOrientation()
        }
    }

    private fun reconcile() {
        for (direction in Camera.Direction.entries) {
            val cameras = cameras.values.filter { it.direction == direction }
            val handleRequired = !cameras.isEmpty()

            if (handleRequired) {
                val handle = cameraHandles.computeIfAbsent(direction) {
                    CameraHandle(
                        plugin,
                        direction,
                        onStateChanged = { updateFlutterCameras(direction) },
                        onRecognitionImage = { image, onComplete ->
                            onRecognitionImage(
                                image,
                                direction,
                                onComplete
                            )
                        }
                    )
                }
                handle.surfaceProducers = cameras.filter { !it.paused }.map { it.surfaceProducer }
            } else {
                cameraHandles.remove(direction)?.close()
            }
            updateFlutterCameras(direction)
        }
    }

    private fun updateFlutterCameras(direction: Camera.Direction) {
        val handle = cameraHandles[direction] ?: return
        val cameras = cameras.values.filter { it.direction == direction && !it.paused }

        var width = handle.size.width
        var height = handle.size.height
        if (handle.quarterTurns % 2 == 1) {
            width = height.also { height = width }
        }

        for (camera in cameras) {
            Handler(Looper.getMainLooper()).post {
                plugin.channel.invokeMethod(
                    "updateCamera",
                    mapOf(
                        "id" to camera.id,
                        "width" to width,
                        "height" to height,
                    )
                )
            }
        }
    }

    private fun onRecognitionImage(
        image: Image,
        direction: Camera.Direction,
        onComplete: () -> Unit
    ) {
        val cameras = cameras.values.filter { it.direction == direction && !it.paused }

        val recognizeText = cameras.any { it.recognizeText }
        val scanBarcodes = cameras.any { it.scanBarcodes }
        val detectFaces = cameras.any { it.detectFaces }

        try {
            ImageRecognition.recognizeImage(
                image,
                recognizeText,
                scanBarcodes,
                detectFaces
            ) { results ->
                val cameras = this.cameras.values.filter { it.direction == direction && !it.paused }
                for (camera in cameras) {
                    Handler(Looper.getMainLooper()).post {
                        plugin.channel.invokeMethod(
                            "recognitionResults",
                            mapOf(
                                "id" to camera.id,
                                "text" to results.text,
                                "barcodes" to results.barcodes,
                                "face" to results.face
                            )
                        )
                    }
                }

                onComplete()
            }
        } catch (_: IllegalStateException) {
            // Ignore
        }
    }
}

File: ios/Classes/Registry.swift
import UIKit

class Registry {
    private let plugin: MulticameraPlugin

    private(set) var cameras: [Int64: Camera] = [:]
    private(set) var cameraHandles: [Camera.Direction: CameraHandle] = [:]

    init(plugin: MulticameraPlugin) {
        self.plugin = plugin
    }

    func registerCamera(
        direction: Camera.Direction,
        paused: Bool,
        recognizeText: Bool,
        scanBarcodes: Bool,
        detectFaces: Bool
    ) -> Int64 {
        let camera = Camera(
            plugin: plugin,
            direction: direction,
            paused: paused,
            recognizeText: recognizeText,
            scanBarcodes: scanBarcodes,
            detectFaces: detectFaces
        )

        cameras[camera.id] = camera
        Task { reconcile() }

        return camera.id
    }

    func updateCamera(
        id: Int64,
        direction: Camera.Direction,
        paused: Bool,
        recognizeText: Bool,
        scanBarcodes: Bool,
        detectFaces: Bool
    ) {
        guard let camera = cameras[id] else { return }

        camera.direction = direction
        camera.paused = paused
        camera.recognizeText = recognizeText
        camera.scanBarcodes = scanBarcodes
        camera.detectFaces = detectFaces

        Task { reconcile() }
    }

    func captureImage(id: Int64, _ callback: @escaping (Data?) -> Void) {
        guard let camera = cameras[id] else {
            callback(nil)
            return
        }
        guard let handle = cameraHandles[camera.direction] else {
            callback(nil)
            return
        }

        handle.captureImage(callback)
    }

    func unregisterCamera(id: Int64) {
        cameras.removeValue(forKey: id)
        Task { reconcile() }
    }

    private func reconcile() {
        for direction in Camera.Direction.allCases {
            let cameras = cameras.values.filter { $0.direction == direction }
            let handleRequired = !cameras.isEmpty

            if handleRequired {
                let handle =
                    cameraHandles[direction]
                    ?? createHandle(direction: direction)

                handle.setCameras(cameras.filter { !$0.paused })
            } else {
                cameraHandles.removeValue(forKey: direction)
            }
            self.updateFlutterCameras(direction)
        }
    }

    private func createHandle(direction: Camera.Direction) -> CameraHandle {
        let handle = CameraHandle(
            direction: direction,
            onCameraUpdated: { [weak self] in
                self?.updateFlutterCameras(direction)
            },
            onRecognitionImage: { [weak self] image in
                self?.onRecognitionImage(image, direction: direction)
            }
        )

        cameraHandles[direction] = handle
        return handle
    }

    private func updateFlutterCameras(_ direction: Camera.Direction) {
        guard let handle = cameraHandles[direction] else { return }
        let cameras = cameras.values.filter { $0.direction == direction }

        var width = handle.size.0
        var height = handle.size.1
        if handle.quarterTurns % 2 == 1 {
            swap(&width, &height)
        }

        for camera in cameras {
            guard let id = camera.id else { continue }
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.01) {
                self.plugin.channel.invokeMethod(
                    "updateCamera",
                    arguments: [
                        "id": id,
                        "width": width,
                        "height": height,
                    ]
                )
            }
        }
    }

    private func onRecognitionImage(
        _ image: UIImage,
        direction: Camera.Direction
    ) {
        let cameras = cameras.values.filter {
            $0.direction == direction && !$0.paused
        }

        let recognizeText = cameras.contains { $0.recognizeText }
        let scanBarcodes = cameras.contains { $0.scanBarcodes }
        let detectFaces = cameras.contains { $0.detectFaces }

        ImageRecognition.recognizeImage(
            image,
            recognizeText: recognizeText,
            scanBarcodes: scanBarcodes,
            detectFaces: detectFaces,
            onResults: { [weak self] results in
                guard let self = self else { return }
                let cameras = self.cameras.values.filter {
                    $0.direction == direction && !$0.paused
                }

                for camera in cameras {
                    guard let id = camera.id else { continue }
                    DispatchQueue.main.async {
                        self.plugin.channel.invokeMethod(
                            "recognitionResults",
                            arguments: [
                                "id": id,
                                "text": results.text as Any,
                                "barcodes": results.barcodes as Any,
                                "face": results.face as Any,
                            ]
                        )
                    }
                }
            }
        )
    }
}
